Dynamo: Support min / max over iterables

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94350
* #94326

Expands support for built-in `min` and `max` calls beyond binary to iterables - simply reduce over the existing binary logic.
Adds support for:
* lists
* tuples
* list iterators
* vararg min / max - `min(2, 3, 4)`

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

jbschlosser committed:
Dynamo: Support min / max over iterables

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -581,7 +581,28 @@ def call_function(
 
-    def _call_min_max(self, tx, a, b):
+    def _call_min_max(self, tx, *args):
+        if len(args) == 1 and args[0].has_unpack_var_sequence(tx):
+            return self._call_min_max_iterable(tx, args[0])
```
Is this ever reached with the `len(items) ==  1` below? or am I misreading it?

voznesenskym commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -581,7 +581,28 @@ def call_function(
+            )
+
+    def _call_min_max_iterable(self, tx, obj):
+        items = obj.unpack_var_sequence(tx)
```
assert has_unpack_var_sequence - I know its checked above, but its good to protect the actual callsite :) it raises NotImplementedError if its not impelemented

ezyang replied:
idk, seems kind of pointless to me, if you've got it wrong unpack_var_sequence will blow up anyway, you aren't getting any safety from an extra assert here

voznesenskym commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -581,7 +581,28 @@ def call_function(
+        elif len(args) == 2:
+            return self._call_min_max_binary(tx, args[0], args[1])
+        else:
+            unimplemented(
+                f"unsupported min / max over {len(args)} args: {[str(a) for a in args]}"
+            )
```
This is a bit redundant with the else / fallthrough case below, doesn't matter too much tho

jbschlosser replied:
True, it's not even that much nicer of a message. I'll remove it

ezyang commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -581,7 +581,28 @@ def call_function(
+        else:
+            unimplemented(
+                f"unsupported min / max over {len(args)} args: {[str(a) for a in args]}"
+            )
```
varargs min is a thing and it doesn't seem like much trouble to support

```
>>> min(1,2,3)
1
```

ezyang commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -581,7 +581,28 @@ def call_function(
+        items = obj.unpack_var_sequence(tx)
+        assert len(items) > 0
+        if len(items) == 1:
+            return items[0]
```
You need to propagate guards from obj to this value!

jbschlosser replied:
It's also done on the return value of the handler:
https://github.com/pytorch/pytorch/blob/fe0e28ab87ddda38a82efe74dd3eaf05e926f117/torch/_dynamo/variables/builtin.py#L563-L565

ezyang reviewed (approved):
Insufficient options propagation

jbschlosser committed:
Update on "Dynamo: Support min / max over iterables"


Expands support for built-in `min` and `max` calls beyond binary to iterables - simply reduce over the existing binary logic.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

jbschlosser committed:
Update on "Dynamo: Support min / max over iterables"


Expands support for built-in `min` and `max` calls beyond binary to iterables - simply reduce over the existing binary logic.
Adds support for:
* lists
* tuples
* list iterators
* vararg min / max - `min(2, 3, 4)`

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge
